cuda_core/cuda/core/experimental/_memory/_buffer.pyx

```diff
@@ -0,0 +1,343 @@
+
+    def copy_from(self, src: Buffer, *, stream: Stream):
+        """Copy from the src buffer to this buffer asynchronously on the given stream.
+
+        Parameters
+        ----------
+        src : :obj:`~_memory.Buffer`
+            Source buffer to copy data from
+        stream : Stream
+            Keyword argument specifying the stream for the
+            asynchronous copy
+
+        """
+        if stream is None:
+            raise ValueError("stream must be provided")
+
+        cdef size_t dst_size = self._size
+        cdef size_t src_size = src._size
+
+        if src_size != dst_size:
+            raise ValueError( "buffer sizes mismatch between src and dst (sizes "
+                             f"are: src={src_size}, dst={dst_size})"
+            )
+        err, = driver.cuMemcpyAsync(self._ptr, src._ptr, dst_size, stream.handle)
+        raise_if_driver_error(err)
+
+    def __dlpack__(
+        self,
+        *,
+        stream: int | None = None,
+        max_version: tuple[int, int] | None = None,
+        dl_device: tuple[int, int] | None = None,
+        copy: bool | None = None,
+    ) -> TypeVar("PyCapsule"):
+        # Note: we ignore the stream argument entirely (as if it is -1).
+        # It is the user's responsibility to maintain stream order.
+        if dl_device is not None:
+            raise BufferError("Sorry, not supported: dl_device other than None")
+        if copy is True:
+            raise BufferError("Sorry, not supported: copy=True")
+        if max_version is None:
+            versioned = False
+        else:
+            if not isinstance(max_version, tuple) or len(max_version) != 2:
+                raise BufferError(f"Expected max_version tuple[int, int], got {max_version}")
+            versioned = max_version >= (1, 0)
+        capsule = make_py_capsule(self, versioned)
+        return capsule
+
+    def __dlpack_device__(self) -> tuple[int, int]:
+        cdef bint d = self.is_device_accessible
+        cdef bint h = self.is_host_accessible
+        if d and (not h):
+            return (DLDeviceType.kDLCUDA, self.device_id)
+        if d and h:
+            # TODO: this can also be kDLCUDAManaged, we need more fine-grained checks
+            return (DLDeviceType.kDLCUDAHost, 0)
+        if (not d) and h:
+            return (DLDeviceType.kDLCPU, 0)
+        raise BufferError("buffer is neither device-accessible nor host-accessible")
+
+    def __buffer__(self, flags: int, /) -> memoryview:
+        # Support for Python-level buffer protocol as per PEP 688.
+        # This raises a BufferError unless:
+        #   1. Python is 3.12+
+        #   2. This Buffer object is host accessible
+        raise NotImplementedError("WIP: Buffer.__buffer__ hasn't been implemented yet.")
+
+    def __release_buffer__(self, buffer: memoryview, /):
+        # Supporting method paired with __buffer__.
+        raise NotImplementedError("WIP: Buffer.__release_buffer__ hasn't been implemented yet.")
```

Comment:
Did we intend to implement this before we merge?

Reply:
it has been there for some time, let's keep this refactoring PR simple